### ISSUE TITLE
Rule to enforce airbnb dangling parameter/argument style

### ISSUE THREAD

**Issue by graingert** [eslint/eslint#7521]
**Please describe what the rule should do:**
Enforce dangling argument and parameter style in functions

**What category of rule is this? **

- [x] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
    // bad
    function foo(bar,
                 baz,
                 quux) {
      // body
    }

    // good
    function foo(
      bar,
      baz,
      quux,
    ) {
      // body
    }

    // bad
    console.log(foo,
      bar,
      baz);

    // good
    console.log(
      foo,
      bar,
      baz,
    );
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It's a pretty simple style and doesn't address a specific / esoteric part of JS

**Comment by graingert** [comment]
aribnb now has this in their styleguide: https://github.com/airbnb/javascript/commit/c51251d17b8e38f1e8975a5e4eb13d7986229410

**Comment by not-an-aardvark** [comment]
Thanks for the suggestion. I have a few questions about the proposal:

How would the rule handle cases like this?

```js
// Are single-line argument lists allowed?
foo(bar, baz);

// Do all arguments need to be on their own line, or just the first argument?
foo(
  bar, baz,
  qux
);

// Does this apply to both function declarations/expressions and call expressions?
// Is there a way to configure it to only apply to one or the other?
function foo(
  bar,
  baz
) {}
foo(
  bar,
  baz
);

```

**Comment by graingert** [comment]
```js
// good
foo(bar, baz);
```
```js
// Do all arguments need to be on their own line, or just the first argument?
foo(
  bar, baz,
  qux
);
```
Probably not, @ljharb confirm?

```js
// Does this apply to both function declarations/expressions and call expressions?
// Is there a way to configure it to only apply to one or the other?
function foo(
  bar,
  baz
) {}
foo(
  bar,
  baz
);
```
should be in an unrelated rule:

```
foo(ham,
    spam)
```

**Comment by graingert** [comment]
It would be great if we could enforce
```js
function foo(bar, baz) { ... }
```
over

```js
function foo(
  bar,
  baz,
) { ... }
```

when there's enough space on the line

**Comment by ljharb** [comment]
In every case, when the `(`/`[`/`{`, `)`/`]`/`}`, and everything between them fit on one line - it's allowed. When they do not, the `(`/`[`/`{` should end a line, and `)`/`]`/`}` should begin a line (at the same indentation level of the beginning of the line that the opening mark began on), and every item should be on its own line by itself, with a trailing comma (which comma-dangle handles), indented one level in from the closing mark (and the line that begins the ending mark).

**Comment by platinumazure** [comment]
@ljharb Can you define "fit on one line"? Is it important that the rule also warn users who *could* fit everything on one line?

**Comment by ljharb** [comment]
No, airbnb's rules would not require that at this time.

That seems like a useful option, perhaps - but unlike the current behavior of object-curly-spacing (i think), we would want to force multiline for > n props, but would want to *separately* require (or allow, or prohibit) single-line for <= n props.

**Comment by ljharb** [comment]
What I mean with "fit on one line" is, simply, is on one line. line length is handled by `max-len`, and a separate rule or option(s) could handle pushing single-line over multi-line - i just want to enforce that *if* it's single-line, it looks one way, and if it's not single-line, it looks another.

**Comment by graingert** [comment]
I actually want to be able to enforce:

```js
function foo(ham,
            spam,
            eggs) {
  ...
}

**Comment by graingert** [comment]
The one line thing is another issue

**Comment by graingert** [comment]
Here's the fit-on-one-line-rule as a separate issue:
https://github.com/eslint/eslint/issues/7526

**Comment by kaicataldo** [comment]
Closing this since there hasn't been any discussion in quite a while and progress on the issue has stalled. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. 

Feel free to open a new issue with a proposal to continue this discussion if this is something you would like to continue advocating for. Please also remember that you can always create this rule for yourself and distribute it as a plugin.
